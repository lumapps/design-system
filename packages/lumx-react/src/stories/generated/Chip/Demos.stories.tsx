/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/chip/Demos' };

export { App as Choice } from './choice';
export { App as Default } from './default';
export { App as Dismissible } from './dismissible';
export { App as Filter } from './filter';
export { App as Group } from './group';
export { App as Small } from './small';
