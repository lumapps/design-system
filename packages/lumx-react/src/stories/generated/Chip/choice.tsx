/home/gcornut/git/dup-design-system/packages/site-demo/content/product/components/chip/react/choice.tsx