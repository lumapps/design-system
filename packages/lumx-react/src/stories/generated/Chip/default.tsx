/home/gcornut/git/design-system/packages/site-demo/content/product/components/chip/react/default.tsx