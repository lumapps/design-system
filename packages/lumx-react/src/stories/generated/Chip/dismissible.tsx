../../../../../site-demo/content/product/components/chip/react/dismissible.tsx