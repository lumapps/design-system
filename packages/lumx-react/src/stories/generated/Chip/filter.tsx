../../../../../site-demo/content/product/components/chip/react/filter.tsx