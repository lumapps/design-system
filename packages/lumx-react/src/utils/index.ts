/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
