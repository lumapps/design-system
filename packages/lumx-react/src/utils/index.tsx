export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
