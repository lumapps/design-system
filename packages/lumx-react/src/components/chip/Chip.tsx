import { Color, ColorPalette, Size, Theme } from '@lumx/react';
import { useStopPropagation } from '@lumx/react/hooks/useStopPropagation';

import { Comp, GenericProps, getRootClassName, handleBasicClasses, onEnterPressed } from '@lumx/react/utils';

import classNames from 'classnames';

import isFunction from 'lodash/isFunction';
import React, { forwardRef, MouseEventHandler, ReactNode } from 'react';

/**
 * Chip sizes.
 */
type ChipSize = Extract<Size, 's' | 'm'>;

/**
 * Defines the props of the component.
 */
export interface ChipProps extends GenericProps {
    /** A component to be rendered after the content. */
    after?: ReactNode;
    /** A component to be rendered before the content. */
    before?: ReactNode;
    /** Color variant. */
    color?: Color;
    /** Whether the component is clickable or not. */
    isClickable?: boolean;
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Whether the chip is currently in a highlighted state or not. */
    isHighlighted?: boolean;
    /** Whether the component is selected or not. */
    isSelected?: boolean;
    /** Size variant. */
    size?: ChipSize;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** On "after" element clicked callback. */
    onAfterClick?: MouseEventHandler;
    /** On "before" element clicked callback. */
    onBeforeClick?: MouseEventHandler;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Chip';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<ChipProps> = {
    size: Size.m,
    theme: Theme.light,
};

/**
 * Chip component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Chip: Comp<ChipProps, HTMLAnchorElement> = forwardRef((props, ref) => {
    const {
        after,
        before,
        children,
        className,
        color,
        disabled,
        isClickable,
        isDisabled = disabled,
        isHighlighted,
        isSelected,
        onAfterClick,
        onBeforeClick,
        onClick,
        size,
        theme,
        ...forwardedProps
    } = props;
    const hasAfterClick = isFunction(onAfterClick);
    const hasBeforeClick = isFunction(onBeforeClick);
    const hasOnClick = isFunction(onClick);

    // Adapt color to the theme.
    const chipColor = color || (theme === Theme.light ? ColorPalette.dark : ColorPalette.light);

    const handleOnBeforeClick = useStopPropagation(onBeforeClick);
    const handleOnAfterClick = useStopPropagation(onAfterClick);

    return (
        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
        <a
            {...forwardedProps}
            ref={ref}
            className={classNames(
                className,
                handleBasicClasses({
                    clickable: Boolean(hasOnClick) || isClickable,
                    color: chipColor,
                    isDisabled,
                    hasAfter: Boolean(after),
                    hasBefore: Boolean(before),
                    highlighted: Boolean(isHighlighted),
                    prefix: CLASSNAME,
                    selected: Boolean(isSelected),
                    size,
                    unselected: Boolean(!isSelected),
                }),
            )}
            role={hasOnClick ? 'button' : undefined}
            tabIndex={isDisabled || !hasOnClick ? -1 : 0}
            aria-disabled={(hasOnClick && isDisabled) || undefined}
            onClick={hasOnClick ? onClick : undefined}
            onKeyDown={hasOnClick ? onEnterPressed(onClick) : undefined}
        >
            {before && (
                // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
                <div
                    className={classNames(`${CLASSNAME}__before`, {
                        [`${CLASSNAME}__before--is-clickable`]: hasBeforeClick,
                    })}
                    onClick={handleOnBeforeClick}
                >
                    {before}
                </div>
            )}

            <div className={`${CLASSNAME}__label`}>{children}</div>

            {after && (
                // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
                <div
                    className={classNames(`${CLASSNAME}__after`, {
                        [`${CLASSNAME}__after--is-clickable`]: hasAfterClick,
                    })}
                    onClick={handleOnAfterClick}
                >
                    {after}
                </div>
            )}
        </a>
    );
});
Chip.displayName = COMPONENT_NAME;
Chip.className = CLASSNAME;
Chip.defaultProps = DEFAULT_PROPS;
