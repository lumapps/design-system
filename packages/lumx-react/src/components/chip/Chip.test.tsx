import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import React, { ReactElement } from 'react';

import { ColorPalette, Theme } from '@lumx/react';
import { Wrapper } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';
import { Chip, ChipProps } from './Chip';

const CLASSNAME = Chip.className as string;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propOverrides: Partial<ChipProps> = {}, shallowRendering = true) => {
    const props = {
        ...propOverrides,
    };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper = renderer(<Chip {...props} />);

    return {
        after: wrapper.find('.lumx-chip__after') as any,
        before: wrapper.find('.lumx-chip__before') as any,
        props,
        wrapper,
    };
};

describe('<Chip />', () => {
    // 1. Test render via snapshot (default state of component).
    describe('Snapshot', () => {
        it('should render correctly Chip component', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });
    });

    // 2. Test defaultProps value and important props custom values.
    // N/A.

    // 3. Test events.
    describe('Events', () => {
        const mockOnClick = jest.fn();
        const mockOnAfterClick = jest.fn();
        const mockOnBeforeClick = jest.fn();
        const mockClickEvent = {
            stopPropagation: (): boolean => true,
        };
        const clearClickMocks = () => {
            [mockOnClick, mockOnAfterClick, mockOnBeforeClick].forEach((func) => func.mockClear());
        };

        beforeEach(() => clearClickMocks);

        it('should trigger onBeforeClick only when clicking on the "before" element', () => {
            const { after, before, wrapper } = setup({
                after: 'after',
                before: 'before',
                onAfterClick: mockOnAfterClick,
                onBeforeClick: mockOnBeforeClick,
                onClick: mockOnClick,
            });

            wrapper.simulate('click', mockClickEvent);
            expect(mockOnBeforeClick).not.toHaveBeenCalled();

            clearClickMocks();

            after.simulate('click', mockClickEvent);
            expect(mockOnBeforeClick).not.toHaveBeenCalled();

            clearClickMocks();

            before.simulate('click', mockClickEvent);
            expect(mockOnBeforeClick).toHaveBeenCalled();
        });

        it('should trigger onClick when clicking on the label area', () => {
            const { after, before, wrapper } = setup({
                after: 'after',
                before: 'before',
                onAfterClick: mockOnAfterClick,
                onBeforeClick: mockOnBeforeClick,
                onClick: mockOnClick,
            });

            wrapper.simulate('click');
            expect(mockOnClick).toHaveBeenCalled();

            clearClickMocks();

            after.simulate('click');
            expect(mockOnClick).not.toHaveBeenCalled();

            clearClickMocks();

            before.simulate('click');
            expect(mockOnClick).not.toHaveBeenCalled();
        });

        it('should trigger onAfterClick only when clicking on the "after" element', () => {
            const { after, before, wrapper } = setup({
                after: 'after',
                before: 'before',
                onAfterClick: mockOnAfterClick,
                onBeforeClick: mockOnBeforeClick,
                onClick: mockOnClick,
            });

            wrapper.simulate('click', mockClickEvent);
            expect(mockOnAfterClick).not.toHaveBeenCalled();

            clearClickMocks();

            after.simulate('click', mockClickEvent);
            expect(mockOnAfterClick).toHaveBeenCalled();

            clearClickMocks();

            before.simulate('click', mockClickEvent);
            expect(mockOnAfterClick).not.toHaveBeenCalled();
        });

        it('should not stop propagation when clicking on a "before"or "after" element without an event handler', () => {
            const onClick = jest.fn();
            const { after, before } = setup(
                {
                    after: 'after',
                    before: 'before',
                    onClick,
                },
                false,
            );

            before.simulate('click', mockClickEvent);
            expect(onClick).toHaveBeenCalled();

            onClick.mockClear();

            after.simulate('click', mockClickEvent);
            expect(onClick).toHaveBeenCalled();
        });
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        it('should have an extra class on the "before" element if it is clickable', () => {
            let { before } = setup({ before: 'before 1' });
            expect(before).toHaveLength(1);
            expect(before.hasClass('lumx-chip__before--is-clickable')).toEqual(false);

            ({ before } = setup({ before: 'before 2', onBeforeClick: (): boolean => true }));
            expect(before).toHaveLength(1);
            expect(before.hasClass('lumx-chip__before--is-clickable')).toEqual(true);
        });

        it('should have an extra class on the "after" element if it is clickable', () => {
            let { after } = setup({ after: 'after 1' });
            expect(after).toHaveLength(1);
            expect(after.hasClass('lumx-chip__after--is-clickable')).toEqual(false);

            ({ after } = setup({ after: 'after 2', onAfterClick: (): boolean => true }));
            expect(after).toHaveLength(1);
            expect(after.hasClass('lumx-chip__after--is-clickable')).toEqual(true);
        });

        it('should have correct default color', () => {
            const { wrapper } = setup({});
            expect(wrapper).toHaveClassName(
                getBasicClass({
                    prefix: CLASSNAME,
                    type: 'color',
                    value: ColorPalette.dark,
                }),
            );
        });

        it('should switch color with theme', () => {
            const { wrapper } = setup({ theme: Theme.dark });
            expect(wrapper).toHaveClassName(
                getBasicClass({
                    prefix: CLASSNAME,
                    type: 'color',
                    value: ColorPalette.light,
                }),
            );
        });

        it('should use color over the theme', () => {
            const color = ColorPalette.red;
            const { wrapper } = setup({ theme: Theme.dark, color });
            expect(wrapper).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'color', value: color }));
        });
    });

    // 5. Test state.
    // N/A.
});
