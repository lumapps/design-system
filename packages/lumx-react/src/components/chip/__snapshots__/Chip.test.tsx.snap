// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chip /> Snapshot should render correctly Chip component 1`] = `
<a
  className="lumx-chip lumx-chip--color-dark lumx-chip--size-m lumx-chip--is-unselected"
  tabIndex={-1}
>
  <div
    className="lumx-chip__label"
  />
</a>
`;
